### ISSUE TITLE
TransAct configuration problem (training issues of AUC=0.5)

### ISSUE THREAD

**Issue by zhujiem** [pinterest/transformer_user_action#1]
Hi, Thanks for the open source code of TransAct!

Recently, I have integreted TransAct to [our FuxiCTR library](https://github.com/reczoo/FuxiCTR/tree/main/model_zoo/TransAct). But I found some confusing results when applying grid search on the hyper-parameters. That is, for some configurations, the model outputs AUC=0.5. It is especially common when setting `k_cols > 1` and `concat_max_pooling=True`. Do you have this problem in your experiments?

For example:
When applying grid search on the following hyperparameters, we get six settings.
```
    transformer_layers: [1, 2, 3]
    k_cols: [1, 2]
    concat_max_pooling: True
```
We obtain the following results, where only the first setting (transformer_layers=1, k_cols=1, and concat_max_pooling=True) produce reasonable result.
```
[test] gAUC: 0.694292 - AUC: 0.695478 -   logloss: 0.425579
[test] gAUC: 0.516050 - AUC: 0.515230 -   logloss: 2.712645
[test] gAUC: 0.500612 - AUC: 0.500910 -   logloss: 2.712645
[test] gAUC: 0.500153 - AUC: 0.500072 -   logloss: 2.712645
[test] gAUC: 0.500000 - AUC: 0.500000 -   logloss: 2.712645
[test] gAUC: 0.499978 - AUC: 0.499977 -   logloss: 2.712645
```

When I set `concat_max_pooling=False`, and keep the other hyperparameters the same:
```
    transformer_layers: [1, 2, 3]
    k_cols: [1, 2]
    concat_max_pooling: False
```
The results become normal in all cases.
```
[test] gAUC: 0.693382 - AUC: 0.693303 - logloss: 0.431847
[test] gAUC: 0.691872 - AUC: 0.695574 - logloss: 0.424100
[test] gAUC: 0.690954 - AUC: 0.686386 - logloss: 0.429123
[test] gAUC: 0.688613 - AUC: 0.686696 - logloss: 0.435287
[test] gAUC: 0.685512 - AUC: 0.687692 - logloss: 0.436102
[test] gAUC: 0.685075 - AUC: 0.688416 - logloss: 0.423795
```

**Comment by GraceJiang0312** [comment]
Hi, I read through the `transact.py` implementation and think I found the root cause.

In `forward()`, when `concat_max_pool=True`, the pooling step is:

```python
pooled_out = self.out_linear(tfmr_out.max(dim=1).values)
```

This takes the max over all sequence positions, including padded ones. `key_padding_mask` only prevents padded positions from being attended to by other positions. It does not zero out the transformer's own output at those padded positions. Since padded positions still pass through the residual connections and FFN, they can produce non-zero output values that are larger in magnitude than the values at real action positions. `max(dim=1)` can then end up selecting values from padding rather than from real user actions.

This would also explain why more transformer layers make it worse: each additional layer can amplify the anomalous padding-position activations further through the FFN, making it increasingly likely that max pooling picks up padding noise instead of signal. That matches a collapse to a constant output, which is consistent with AUC landing at 0.5.

This is also consistent with `concat_max_pool=False` working fine across all configurations, since removing that branch removes the only place where padded positions' outputs get selected without being masked.

A possible fix: mask out padded positions' output (e.g., fill with `-inf`) before the `max(dim=1)` call, similar to how the attention mask already excludes them from attention.

I have not run this against real data to confirm, so treat this as a code-read hypothesis rather than a verified fix.
